enhancements/network/egress-ip-per-destinations.md

```diff
@@ -0,0 +1,327 @@
+- "@kyrtapz"
+approvers:
+- "@trozet"
+api-approvers:
+- "@joelspeed"
+creation-date: 2024-03-12
+last-updated: 2024-03-12
+tracking-link:
+- "https://issues.redhat.com/browse/SDN-4454"
+see-also:
+- "NA"
+replaces:
+- "NA"
+superseded-by:
+- "NA"
+---
+
+# Egress IP per destination
+## Summary
+Today, we can use an `EgressIP` to describe the source IP for a pod if it selected by an `EgressIP` custom resource (CR) selectors. This includes namespace
+and pod selectors. If multiple `EgressIP` CRs select the same set of pods, the behavior is undefined. This is because
+we cannot reliably choose which source IP to use.
+
+This enhancement proposes adding a new selector for when we want to apply the `EgressIP`. This new selector will only apply
+the `EgressIP` as the source IP to a set of pods communicating with destination IP if that destination IP is within predefined network CIDRs.
+
+If the new destination traffic selector is specified for all `EgressIP` CRs that selected a set of pods and the destination networks selected do not overlap, then we
+can allow a set of pods to have multiple source IPs depending on the destination address.
+
+## Motivation
+### User Stories
+
+As a cluster tenant, I want my application traffic to egress a particular interface depending on the traffic destination address and must
+have a distinct but consistent source IP for each interface.
+
+As a cluster administrator, I want to easily configure which destination networks should have an `EgressIP` because the destination
+networks may change, and I do not wish to have to update individual `EgressIP` custom resources for every update.
+
+### Goals
+- Allow a set of pods to have a consistent source IP depending on the destination
+- Support shared and local gateway (aka `routingViaHost=true`) modes in-order to reduce asymmetry between the modes
+- Backward compatibility with existing `EgressIP` API
+- If the new API isn't defined, preserve existing behaviour
+
+### Non-Goals
+- Change existing `EgressIP` behavior when the new field is not defined
+
+## Proposal
+
+A new `TrafficSelector` field is added to `EgressIP` spec which uses Kubernetes label selector semantics
+to select a new CRD called `EgressIPTraffic` which contains a list of destination networks. This list of networks defines when an `EgressIP` should
+be used as source IP when an application attempts to talk to a destination address, and it is within one of the predefined networks
+defined within a `EgressIPTraffic` CR.
+
+The reason for not embedding the destination networks within each `EgressIP` CR is that it places a burden
+on the cluster administrator to update all `EgressIP` CRs if the destination networks change.
+
+With the Kubernetes label selector semantics, cluster administrators can define a set of destination networks that multiple `EgressIP`
+CRs can consume.
+
+If greater than one `EgressIPTraffic` CRs are selected, the networks are added together, and if a destination IP is within one
```

Comment:
If they have overlapping subnets, its not a concern because all we care about is if the destination IP is within one of the subnets. Once one matches, SNATing occurs.

Lets say we have `EIPTraffic1` with networks `1.1.1.0/24` and `EIPTraffic2` with networks `1.1.0.0/16`. The aforementioned `EgressIPTraffic` CRs are selected by an EIP and if a destination IP falls within either, then SNAT occurs.

Reply:
```
If a pod attempts to connect to an IP in a network that is selected by multiple `EgressIP`, then the behaviour is undefined, and either `EgressIP` maybe selected as source IP.
```

Added this to the enhancement.